process/process.go

```diff
@@ -1,145 +1,187 @@
 	}
 
-	// Provision non-provisoned online devices associated with this application
-	if errs := a.ProvisionDevices(); errs != nil {
-		return errs
+	if log.GetLevel() == log.DebugLevel {
+		log.WithFields(log.Fields{
+			"Provisioned devices": provisionedDevices,
+		}).Debug("Processing application")
+	} else {
+		log.WithFields(log.Fields{
+			"Number of provisioned devices": len(provisionedDevices),
+		}).Info("Processing application")
 	}
 
-	// Set the status of all offline provisioned devices associated with this application to OFFLINE
-	if errs := a.SetOfflineDeviceStatus(); errs != nil {
-		return errs
+	// Convert provisioned devices to a hash map
+	hashmap := make(map[string]struct{})
+	var s struct{}
+	for _, value := range provisionedDevices {
+		hashmap[value.LocalUUID] = s
 	}
 
-	// Update firmware for all online devices associated with this application
-	if errs := a.UpdateOnlineDevices(); errs != nil {
-		return errs
+	// Provision all unprovisioned devices associated with this application
+	for key := range onlineDevices {
+		if _, ok := hashmap[key]; ok {
+			// Device already provisioned
+			continue
+		}
+
+		// Device not already provisioned
+		if errs := provisionDevice(a, key); errs != nil {
+			return errs
+		}
 	}
 
-	// Update config for all online devices associated with this application
-	if errs := a.UpdateConfigOnlineDevices(); errs != nil {
-		return errs
+	// Refesh all provisioned devices associated with this application
+	provisionedDevices, err = getProvisionedDevices(a)
+	if err != nil {
+		return []error{err}
 	}
 
-	// Update environment for all online devices associated with this application
-	if errs := a.UpdateEnvironmentOnlineDevices(); errs != nil {
-		return errs
+	// Sync all provisioned devices associated with this application
+	for _, value := range provisionedDevices {
+		if errs := value.Sync(); errs != nil {
+			return errs
+		}
+
+		if err := updateDevice(value); err != nil {
+			return []error{err}
+		}
 	}
 
-	// Handle device flags
-	if err := a.HandleFlags(); err != nil {
+	// Refesh all provisioned devices associated with this application
+	provisionedDevices, err = getProvisionedDevices(a)
+	if err != nil {
 		return []error{err}
 	}
 
-	return nil
-}
+	// Set state for all provisioned devices associated with this application
+	for _, value := range provisionedDevices {
+		if _, ok := onlineDevices[value.LocalUUID]; ok {
+			value.Status = deviceStatus.IDLE
+		} else {
+			value.Status = deviceStatus.OFFLINE
+		}
+
+		if err := updateDevice(value); err != nil {
+			return []error{err}
+		}
 
-func Pending() {
-	for _, a := range application.List {
-		for _, d := range a.Devices {
-			if d.Commit != d.TargetCommit && d.Status != deviceStatus.OFFLINE {
-				UpdatesPending = true
-				return
+		if errs := sendState(value); errs != nil {
+			return errs
+		}
+	}
+
+	// Refesh all provisioned devices associated with this application
+	provisionedDevices, err = getProvisionedDevices(a)
```

Comment:
The aim with writing to the database every step is avoid data loss if the container or device is restarted. The aim with reading from the database every step is to ensure I pick up changes written to the database by the `api/handlers` as soon as possible.